sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -981,6 +982,10 @@ object EliminateSorts extends Rule[LogicalPlan] {
       j.copy(left = recursiveRemoveSort(originLeft), right = recursiveRemoveSort(originRight))
     case g @ Aggregate(_, aggs, originChild) if isOrderIrrelevantAggs(aggs) =>
       g.copy(child = recursiveRemoveSort(originChild))
+    case r: RepartitionByExpression =>
```

Comment:
These two branches can be replaces with one:

```
case r: RepartitionOperation =>
  r.withNewChildren(r.children.map(recursiveRemoveSort))
```

It will mean all repartition nodes we add in the future will be also taken into account. It seems safe but I want to hear what everybody thinks.

Reply:
What about adding `RepartitionOperation.preservesOrder`? Then we could collapse these cases while also excluding `Coalesce` and making this explicit for future repartition operators.

Reply:
I am +1 on having a repartition node that preserves ordering. In fact, we have such a node internally. `Coalesce` is not really order preserving, though. It has custom logic in `DefaultPartitionCoalescer` that gets applied if the parent RDD has locality info. We would also need to report `outputOrdering` correctly (which is not done in case of `Coalesce` now)

Reply:
We won't be able to squash all cases into one as we need to check if repartition expressions are deterministic. However, I'd consider extending repartition nodes with order preserving repartition in a follow-up if there is enough support for that.